buildSrc/src/main/java/org/elasticsearch/gradle/internal/test/rerun/executer/RerunTestResultProcessor.java

```diff
@@ -0,0 +1,121 @@
+
+final class RerunTestResultProcessor implements TestResultProcessor {
+
+    private final TestResultProcessor delegate;
+
+    private final Map<Object, TestDescriptorInternal> activeDescriptorsById = new HashMap<>();
+
+    private TestNames currentRoundFailedTests = new TestNames();
+    private TestNames previousRoundFailedTests = new TestNames();
+
+    private Object rootTestDescriptorId;
+
+    private List<Map<Object,TestDescriptorInternal>> storedActiveDescriptorsWhenFailed = new ArrayList<>();
+
+    RerunTestResultProcessor(TestResultProcessor delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public void started(TestDescriptorInternal descriptor, TestStartEvent testStartEvent) {
+        if (rootTestDescriptorId == null) {
+            rootTestDescriptorId = descriptor.getId();
+            activeDescriptorsById.put(descriptor.getId(), descriptor);
+            delegate.started(descriptor, testStartEvent);
+        } else if (descriptor.getId().equals(rootTestDescriptorId) == false) {
+            activeDescriptorsById.put(descriptor.getId(), descriptor);
+            delegate.started(descriptor, testStartEvent);
+        }
+    }
+
+    @Override
+    public void completed(Object testId, TestCompleteEvent testCompleteEvent) {
+        if (testId.equals(rootTestDescriptorId)) {
+            if (lastRun() == false) {
+                return;
+            }
+        } else {
+            TestDescriptorInternal descriptor = activeDescriptorsById.remove(testId);
+            if (descriptor != null && descriptor.getClassName() != null) {
+                String className = descriptor.getClassName();
+                String name = descriptor.getName();
+                boolean failedInPreviousRound = previousRoundFailedTests.remove(className, name);
+                if (failedInPreviousRound && testCompleteEvent.getResultType() == SKIPPED) {
+                    currentRoundFailedTests.add(className, name);
+                }
+            }
+        }
+
+        delegate.completed(testId, testCompleteEvent);
+    }
+
+    @Override
+    public void output(Object testId, TestOutputEvent testOutputEvent) {
+        delegate.output(testId, testOutputEvent);
+    }
+
+    @Override
+    public void failure(Object testId, Throwable throwable) {
+        final TestDescriptorInternal descriptor = activeDescriptorsById.get(testId);
+        if (descriptor != null) {
+            String className = descriptor.getClassName();
+            if (className != null) {
+                currentRoundFailedTests.add(className, descriptor.getName());
+            }
+        }
+
+        delegate.failure(testId, throwable);
+    }
+
+    private boolean lastRun() {
+        return activeDescriptorsById.size() == 1;
+    }
+
+    public RoundResult getResult() {
+        return new RoundResult(currentRoundFailedTests, previousRoundFailedTests, lastRun());
+    }
+
+    public void reset(boolean lastRetry) {
```

Comment:
Looks like the `lastRetry` argument isn't used.

Reply:
cleaned up